src/backends/plonky2/primitives/merkletree/circuit.rs

```diff
@@ -396,105 +398,136 @@ fn kv_hash_target(
 /// Verifies that the merkletree state transition (from old_root to new_root)
 /// has been done correctly for the given new_key. This will allow verifying
 /// correct new leaf insertion, and leaf edition&deletion (if needed).
-pub struct MerkleProofStateTransitionTarget {
+/// See `MerkleTreeStateTransitionProof` struct for an explanation of the fields.
+pub struct MerkleTreeStateTransitionProofTarget {
     pub(crate) max_depth: usize,
     // `enabled` determines if the merkleproof state transition verification is enabled
     pub(crate) enabled: BoolTarget,
-    pub(crate) typ: Target,
-
+    pub(crate) op: Target,
     pub(crate) old_root: HashOutTarget,
-    pub(crate) proof_non_existence: MerkleClaimAndProofTarget,
-
+    pub(crate) op_proof: MerkleClaimAndProofTarget,
     pub(crate) new_root: HashOutTarget,
-    pub(crate) new_key: ValueTarget,
-    pub(crate) new_value: ValueTarget,
-    pub(crate) new_siblings: Vec<HashOutTarget>,
+    pub(crate) op_key: ValueTarget,
+    pub(crate) op_value: ValueTarget,
+    pub(crate) siblings: Vec<HashOutTarget>,
 
     // auxiliary witness
     pub(crate) divergence_level: Target,
 }
 /// creates the targets and defines the logic of the circuit
 pub fn verify_merkle_state_transition_circuit(
     builder: &mut CircuitBuilder<F, D>,
-    proof: &MerkleProofStateTransitionTarget,
+    proof: &MerkleTreeStateTransitionProofTarget,
 ) {
     let measure = measure_gates_begin!(
         builder,
         format!("MerkleProofStateTransition_{}", proof.max_depth)
     );
     let zero = builder.zero();
-
-    // for now, only type=0 (insertion proof) is supported
-    builder.connect(proof.typ, zero);
-
-    // 1) check that for the old_root, the new_key does not exist in the tree
-    verify_merkle_proof_circuit(builder, &proof.proof_non_existence);
-
-    // 2) check that for the new_root, the new_key does exist in the tree
+    let one = builder.one();
+    let two = builder.constant(F::from_canonical_u64(2));
+
+    // Op type check
+    let is_insertion = builder.is_equal(proof.op, zero);
+    let is_update = builder.is_equal(proof.op, one);
+    let is_deletion = builder.is_equal(proof.op, two);
+    let op_type_check = {
+        let a = builder.or(is_insertion, is_update);
+        builder.or(a, is_deletion)
+    };
+    builder.assert_one(op_type_check.target);
+
+    // 1) Verify the provided op proof.
+    verify_merkle_proof_circuit(builder, &proof.op_proof);
+
+    // 2) Check that the provided siblings yield an existence proof of
+    // (op_key, op_value) with root specified by op:
+    // Insert: `new_root`
+    // Update: `new_root`
+    // Delete: `old_root`
+    let root = HashOutTarget {
```

Comment:
in `src/backends/plonky2/circuits/common.rs` we have the method `select_flattenable` and also `impl Flattenable for HashOutTarget`.
This would allow implementing this as:
```rust
let root = builder.select_flattenable(is_deletion, proof.old_root, proof.new_root)
```

I understand you may not want to import `common` because it has many pod-isms.  Would it make sense to split `common` into pod-isms and non-pod-isms to be used around?